### ISSUE TITLE
Segfault when running in daemon mode

### ISSUE THREAD

**Issue by anonimal** [monero-project/kovri#312]
**By submitting this issue, I confirm the following:**
- I have read and understood the [contributor guide](https://github.com/monero-project/kovri/blob/master/doc/CONTRIBUTING.md).
- I have checked that the issue I am reporting can be replicated or that the feature I am suggesting is not present.
- I have checked opened or recently closed [pull requests](https://github.com/monero-project/kovri/pulls) for existing solutions/implementations to my issue/suggestion.

_Place an X inside the bracket to confirm_
- [X] I confirm.

---

Tested + reproducible on:
- Ubuntu 16.04, (gcc 5.4.0)
- Arch Linux (gcc 6.1.1)
- OSX (Apple LLVM version 6.0 (clang-600.0.57) (based on LLVM 3.5svn))

Running`./build/kovri/ -d1` with regular `make` against b310b19.

GDB backtrace [attached](https://github.com/monero-project/kovri/files/422486/daemon_backtrace.txt).

OSX produces (did not run LLDB):

```
libc++abi.dylib: terminating with uncaught exception of type std::__1::system_error: mutex lock failed: Invalid argument
Abort trap: 6
```

Daemon mode was fully functional when I opened #214 but building against b078bd5 produces the same results.

**Edit**: side-note and possibly not related: there have been multiple issues in the past when in daemon mode and the daemon-related code is just terrible - as is the singleton tied into it.

**Comment by anonimal** [comment]
Related to #342. First 10 frames are identical to OP backtrace. Issue could be a combination of client and threading.

Note: fe270c0 does not fix this issue.

**Comment by anonimal** [comment]
> Daemon mode was fully functional when I opened #214 but building against b078bd5 produces the same results.

I don't know what was going on there, I could've been mistaken or had confused one running instance for another. This problem was introduced 2f3903407a42c95c766c050e244bc8ae3ebd6e28 (confirmed against 6efa1439c20bb274539cb218d117417536ebb891). Fixed in #484.
